design-proposals/network-vhostuser-binding-plugin/network-vhostuser-binding-plugin.md

```diff
@@ -0,0 +1,141 @@
+            <cell id="1" cpus="8-15" memory="2097152" unit="KiB" memAccess="shared"/>
+        </numa>
+</cpu>
+<interface type='vhostuser'>
+    <source type='unix' path='/var/run/kubevirt/sockets/poda08a0fcbdea' mode='server'/>
+    <target dev='poda08a0fcbdea'/>
+    <model type='virtio-non-transitional'/>
+    <mac address='ca:fe:ca:fe:42:42'/>
+    <driver name='vhost' queues='8' rx_queue_size='1024' tx_queue_size='1024'/>
+    <alias name='ua-net1'/>
+</interface>
+```
+
+This design leverages the existing `sockets` emptyDir mounted in `/var/run/kubevirt/sockets`. This allows the CNI to bind mount the socket emptyDir (`/var/lib/kubelet/<pod uid>/volumes/kubernetes.io~empty-dir/sockets`) to a host directory available to the dataplane pod through a hostPath mount.
+
+However this assumes:
+- the `sockets` emptyDir can be used for such a purpose
+- the CNI has access to the `sockets` emptyDir of the `virt-launcher` pod and that it can bind mount it to a path available to the dataplane. This can be tricky especially with Multus 4 in thick plugin mode, where the CNI is executed by the mutlus pod. Usually the multus thick plugin daemonset defines a `/hostroot` hostPath volume mount with `mountPropagation: HostToContainer` option. A `mountPropagation: Bidirectional` option is needed for the bind mount to be propagated back to the host and to the dataplane pod.
+- the dataplane pod has the privilege to mount a hostPath
+
+Here is a diagram showing sockets sharing mecanisms between `virt-launcher` pod, `userspace CNI` and the `dataplane` pod.
+
+![kubevirt-vhostuser-shared-sockets](kubevirt-vhostuser-shared-sockets.png)
+
+Sharing the `vhostuser` sockets between `virt-launcher` pods and the dataplane pod is something to be enhanced in order to limit usage of hostPath volumes and bind mounts.
+
+## Alternative designs
+
+Some alternative designs were discussed in ![kubevirt-dev mailing-list](https://groups.google.com/g/kubevirt-dev/c/3w_WStrJfZw/m/yWSBpDAKAQAJ).
+
+### Expose a virt-launcher pod directory to binding plugin
+
+This requires to implement a new network binding plugin mechanism we could expose the content of a `virt-launcher` directory to an external plugin.
+The plugin registration in Kubervirt resource would define the target directory on the node where the directory should be exposed.
+
+This diagram explains this mechanism.
+
+![kubevirt-plugin-extension](kubevirt-plugin-extension.png)
+
+The advantages of such an approach are:
+- is generic and could also be potentially reused by other device types
+- hides the KubeVirt implementation details. Currently, you need to know where the KubeVirt sockets are located in the virt-launcher filesystem. Potentially, if we change the directory path for the sockets, this would break the CNI plugin
+- can isolate the resources dedicated to that particular plugin
+
+The drawback side is that `virt-launcher` pod would need to be run with `privileged: true` in order to do the bind mount.
```

Comment:
I don't think this is true. The directory can be exposed by virt-handler. Hence, no need of privileged for virt-launcher. This would also be unfeasible since virt-launcher is untrusted

Reply:
I don't see very well the interaction between virt-handler, virt-launcher and the exposed directories. Need to check that.